File: src/main/java/com/owant/thinkmap/model/NodeModel.java
package com.owant.thinkmap.model;

import java.io.Serializable;
import java.util.LinkedList;

/**
 * Created by owant on 09/02/2017.
 */
public class NodeModel<T> implements Serializable {


    /**
     * the parent node,if root node parent node=null;
     */
    public NodeModel<T> parentNode;

    /**
     * the data value
     */
    public T value;

    /**
     * have the child nodes
     */
    public LinkedList<NodeModel<T>> childNodes;

    /**
     * focus tag for the tree add nodes
     */
    public transient boolean focus = false;

    /**
     * index of the tree floor
     */
    public int floor;

    public boolean hidden = false;

    public NodeModel(T value) {
        this.value = value;
        this.childNodes = new LinkedList<>();

        this.focus = false;
        this.parentNode = null;
    }

    public NodeModel<T> getParentNode() {
        return parentNode;
    }

    public void setParentNode(NodeModel<T> parentNode) {
        this.parentNode = parentNode;
    }

    public T getValue() {
        return value;
    }

    public void setValue(T value) {
        this.value = value;
    }

    public LinkedList<NodeModel<T>> getChildNodes() {
        return childNodes;
    }

    public void setChildNodes(LinkedList<NodeModel<T>> childNodes) {
        this.childNodes = childNodes;
    }

    public boolean isFocus() {
        return focus;
    }

    public void setFocus(boolean focus) {
        this.focus = focus;
    }

    public int getFloor() {
        return floor;
    }

    public void setFloor(int floor) {
        this.floor = floor;
    }

    public boolean isHidden() {
        return hidden;
    }

    public void setHidden(boolean hidden) {
        this.hidden = hidden;
    }

}

File: library/src/main/java/com/owant/thinkmap/NodeModel.kt
package com.owant.thinkmap

import java.util.*

/**
 *  created by Kyle.Zhong on 2020/6/8
 *  节点
 */
open class NodeModel<T> {

    /**
     * 父节点
     */
    var parentNode: NodeModel<T>? = null

    /**
     * 当前节点的数据
     */
    var nodeValue: T

    /**
     * 子节点集合
     */
    private var childNodes: LinkedList<NodeModel<T>> = LinkedList<NodeModel<T>>()

    /**
     * 当前的层级
     */
    var floor: Int = 0

    /**
     * 是否关闭
     */
    var close: Boolean? = false

    /**
     * 是否对焦
     */
    var focusing: Boolean? = false


    fun addNode(child: NodeModel<T>) {
        child.parentNode = this
        child.floor = floor + 1
        getNodes().add(child)
    }

    fun getNodes(): LinkedList<NodeModel<T>> {
        return childNodes
    }

    constructor(nodeValue: T) {
        this.nodeValue = nodeValue
    }

    override fun toString(): String {
        return "NodeModel(parentNode not empty=${parentNode != null}, nodeValue=${nodeValue.toString()}, childSize=${childNodes.size}, floor=$floor, close=$close, focusing=$focusing)"
    }


}
